src/test/java/com/google/devtools/build/lib/rules/cpp/LibrariesToLinkCollectorTest.java

```diff
@@ -0,0 +1,246 @@
+import com.google.devtools.build.lib.packages.util.Crosstool.CcToolchainConfig;
+import com.google.devtools.build.lib.packages.util.ResourceLoader;
+import com.google.devtools.build.lib.testutil.TestConstants;
+import com.google.devtools.build.lib.vfs.FileSystemUtils;
+import java.util.List;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class LibrariesToLinkCollectorTest extends BuildViewTestCase {
+
+  @Test
+  public void getRalitive_returnsRelativePaths() {
+    assertThat(getRelative(create("foo"), create("foo/bar/baz"))).isEqualTo(create("bar/baz"));
+    assertThat(getRelative(create("/foo"), create("/foo/bar/baz"))).isEqualTo(create("bar/baz"));
+    assertThat(getRelative(create("foo/bar"), create("foo/bar/baz"))).isEqualTo(create("baz"));
+    assertThat(getRelative(create("/foo/bar"), create("/foo/bar/baz"))).isEqualTo(create("baz"));
+    assertThat(getRelative(create("/"), create("/foo"))).isEqualTo(create("foo"));
+    assertThat(getRelative(create(""), create("foo"))).isEqualTo(create("foo"));
+    assertThat(getRelative(create(""), create("foo/bar"))).isEqualTo(create("foo/bar"));
+    assertThat(getRelative(create("/foo/bar"), create("/foo/bar")).getPathString()).isEmpty();
+
+    assertThat(getRelative(create("foo/bar/baz"), create("foo"))).isEqualTo(create("../.."));
+    assertThat(getRelative(create("/foo/bar/baz"), create("/foo"))).isEqualTo(create("../.."));
+    assertThat(getRelative(create("foo/bar/baz"), create("foo/bar"))).isEqualTo(create(".."));
+    assertThat(getRelative(create("/foo/bar/baz"), create("/foo/bar"))).isEqualTo(create(".."));
+    assertThat(getRelative(create("/foo"), create("/"))).isEqualTo(create(".."));
+    assertThat(getRelative(create("foo"), create(""))).isEqualTo(create(".."));
+    assertThat(getRelative(create("foo/bar"), create(""))).isEqualTo(create("../.."));
+    assertThat(getRelative(create("/foo/baz"), create("/foo/bar"))).isEqualTo(create("../bar"));
+    assertThat(getRelative(create("bar"), create("foo"))).isEqualTo(create("../foo"));
+
+    assertThat(getRelative(create("foo"), create("../foo"))).isEqualTo(create("../../foo"));
+    assertThat(getRelative(create(".."), create("../../foo"))).isEqualTo(create("../foo"));
+    assertThat(getRelative(create("../bar"), create("../../foo"))).isEqualTo(create("../../foo"));
+  }
+
+  @Test
+  public void getRelative_throwsOnInvalidCases() {
+    assertThrows(IllegalArgumentException.class, () -> getRelative(create("bar"), create("/foo")));
+    assertThrows(IllegalArgumentException.class, () -> getRelative(create("/bar"), create("foo")));
+    assertThrows(IllegalArgumentException.class, () -> getRelative(create(".."), create("")));
+    assertThrows(
+        IllegalArgumentException.class, () -> getRelative(create("../../bar"), create("../foo")));
+  }
+
+  @Test
+  public void dynamicLink_siblingLayout_externalBinary_rpath() throws Exception {
```

Comment:
It would be great to also have an integration test (probably in `cc_integration_test.sh`).

Reply:
Testing toolchain lib rpaths is a bit tricky.

1. I need a `cc_toolchain_config` with `static_link_cpp_runtimes` turned on.
2. I need a `cc_toolchain` that sets the `dynamic/static_runtime_lib` attributes with runtime libs (`libc++` or `libstdc++`).
3. Since an incorrect rpath does not fail the build, I have to run the binary to verify that it can correctly load the libs.

It seems that 1 and 2 are not supported by the auto-generated toolchain, so that I need to somehow replicate the default toolchain / config and do some modifications, right? Or I can create a custom toolchain - but since there's no hermetic compiler available, I'm not sure how far I could get.

3 means that I can't use a fake runtime lib (can I?), or the binary will fail to load as well. What I can think of is searching for them in the host (again since there is no hermetic compiler available) - I definitely don't want to pull llvm as a test dependency here.

Considering all these requirements, I'm not sure if it is worth the effort. Do you have any pointers for how to implement the test?

Reply:
Unless I'm missing something --features=static_link_cpp_runtimes be enough for point 1.

Regarding point 2, would it be useful in the future to have a Starlark label-typed build setting that is able to override the `dynamic_runtime_lib` and the `static_runtime_lib`  values  (i.e. a command line option accepting a file target that overrides those attributes). If that's actually something that would be widely useful, we could implement that once we starlarkify C++ configurations (or we could add it already with a native LateBoundDefault if it's something that would be really really useful).

If such a flag is not worth the complexity because it would only be useful for this test, I still feel like this should be checked in tested though even if it requires mocking a new toolchain.

Reply:
re: `--features=static_link_cpp_runtimes`, it's not working now. It seems that the `static_link_cpp_runtimes` feature is not defined in the [default unix toolchain config](https://cs.opensource.google/bazel/bazel/+/master:tools/cpp/unix_cc_toolchain_config.bzl), nor is it added by `CppActionConfigs.getLegacyFeatures()` because `supportsEmbeddedRuntimes` is [set to false](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/cpp/CcModule.java;drc=1623f7eb7850b6e92010d5f76607c0cb019cb159;l=1231). The relevant commit is ad30d85c64d629ad64ffd85816dd2a067d424c07. Maybe I could add the feature (but disabled) to `unix_cc_toolchain_config.bzl`, but not sure if it would cause undesirable effects.

re: point 2, one workaround would be duplicating the default `cc_toolchain` target (e.g. `@local_config_cc//:cc-compiler-k8`) and setting `dynamic|static_runtime_lib` on the copy (reusing the default `cc_toolchain_config` and dependencies). Not sure how good it works cross-platform, though. A config to override them would be welcome, but I'm not sure how wide it can be useful. ([Here](https://github.com/bazelbuild/bazel/issues/16520#issuecomment-1409252801) are a set of more general ideas that I think would be quite useful).

Would you be OK if I add the `static_link_cpp_runtimes` feature into `unix_cc_toolchain_config.bzl` (disabled by default)? I think it will be a different PR, though.

Reply:
> Would you be OK if I add the static_link_cpp_runtimes feature into unix_cc_toolchain_config.bzl (disabled by default)? I think it will be a different PR, though.

Yes, that is a welcome change if that helps getting you there. If it's disabled by default it should be safe.

> (https://github.com/bazelbuild/bazel/issues/16520#issuecomment-1409252801 are a set of more general ideas that I think would be quite useful).

I haven't spent time fully processing your ideas there but please take a look at
https://docs.google.com/document/d/1-etGNsPneQ8W7MBMxtLloEq-Jj9ng1G-Pip-cWtTg_Y/edit?pli=1#heading=h.5mcn15i0e1ch and tell me if that kind of solves the same problem you are trying to solve with those ideas.

Reply:
https://github.com/bazelbuild/bazel/pull/17391 to add the feature.

Thanks for sharing the proposal! That doc fully covers points 2 and 3 in my comment regarding the toolchain, and mostly covers point 5 (although is a P2). It could potentially also cover point 1 (I left a comment on the proposal). The author mentioned in the document history that there will be another proposal for something similar to point 4 (I hope) - for me point 4 is a killer feature that I'm implementing with some custom rules but still rough due to limitations in the current cc_toolchain / cc_toolchain_config rules.

re: ideas about the build rules, they apparently belong to a separate proposal.

So yeah this proposal definitely overlaps with many of my ideas. Glad and can't wait to see the improvements there!

Reply:
Let me know if I missed something and you can indeed add a test easily already or alternatively let me know if this PR doesn't block you and it can wait for quite a while till the cc_toolchain design is implemented.

If it needs to go in now already to unblock you, please just add the comment and I will approve.

Reply:
Added a TODO item in the test class.

I don't think there is an easy way to do that yet. This PR will block me soon so yeah let's get it merged.

Thank you all for the review @oquenchil @fmeum!